firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,311 @@
+			data_length = length;
+			I2C_GenerateSTART(I2C_config.HWConfig->I2CPeriph, ENABLE);
+			task_delivery = true;
+		}
+	}
+	return task_delivery;
+}
+
+bool I2C_receive(I2C_channel_E channel, const uint8_t * data, const uint8_t length)
+{
+	bool task_delivery = false;
+
+	// Check the validity of message
+	if((data != NULL) && (length != 0) && (length <= _I2C_BUFFER_SIZE))
+	{
+		// Check the availability of I2C
+		if ((data_length == 0) && (I2C_state == I2C_STATE_IDLE)) {
+
+			memcpy(I2C_receive_data.Buffer, data, length);
+			I2C_state = I2C_STATE_RECEIVE;
+			slave_address = I2C_channelToAddressMapping(channel);
+			data_length = length;
+			I2C_GenerateSTART(I2C_config.HWConfig->I2CPeriph, ENABLE);
+			task_delivery = true;
+		}
+	}
+	return task_delivery;
+}
+
+bool is_idle()
+{
+	if (I2C_state != I2C_STATE_IDLE) return false;
+	else return true;
+}
+
+// See stm32f413 programming manual P855-861
+static inline void I2C_private_TxRxIRQ(I2C_TypeDef * I2Cx)
+{
+	// Check Start condition
+	if ((I2Cx->SR1 & I2C_SR1_SB) == I2C_SR1_SB)
+	{
+		// LSB is the r/w bit
+		I2C_Send7bitAddress(I2Cx, (slave_address << I2C_OAR1_ADD0), (uint8_t)I2C_state);
+	}
+	// Waits for address sent bit to be set
+	else if ((I2Cx->SR1 & I2C_SR1_ADDR) == I2C_SR1_ADDR)
+	{
+		// Clearing ADDR bit by reading SR1 and then reading SR2
+		I2Cx->SR2;
+		if (I2C_state == I2C_STATE_SEND)
+		{
+			sendBufferPtr = &I2C_send_data.Buffer[0];
+			I2C_SendData(I2Cx, *sendBufferPtr);
+			sendBufferPtr ++;
+			data_length --;
+		}
+		else if (I2C_state == I2C_STATE_RECEIVE)
+		{
+			receiveBufferPtr = &I2C_receive_data.Buffer[0];
+			if (data_length == 1)
+			{
+				I2C2->CR1 &= ~(I2C_CR1_ACK);
+				I2C_GenerateSTOP(I2Cx, ENABLE); // Generate a stop condition
+			}
+		}
+	}
+	// Sending data to slave
+	else if ((I2Cx->SR1 & I2C_SR1_TXE) == I2C_SR1_TXE)
+	{
+		if (data_length > 0)
+		{
+			I2C_SendData(I2Cx, *sendBufferPtr);
+			sendBufferPtr ++;
+			data_length --;
+		}
+		else
+		{
+			// Generate a stop condition ONLY ONCE
+			if (I2C_state != I2C_STATE_IDLE){
+				I2C_GenerateSTOP(I2Cx, ENABLE);
+			}
+			I2C_state = I2C_STATE_IDLE;
+		}
+	}
+	// Receiving data from slave
+	else if ((I2Cx->SR1 & I2C_SR1_RXNE) == I2C_SR1_RXNE)
+	{
+		*receiveBufferPtr = I2C_ReceiveData(I2Cx);
+		receiveBufferPtr ++;
+		data_length --;
+		if (data_length == 1)
+		{
+			I2Cx->CR1 &= ~(I2C_CR1_ACK);
+			I2C_GenerateSTOP(I2Cx, ENABLE); // Generate a stop condition
+		}
+		if (data_length == 0)
+		{
+			I2C_state = I2C_STATE_IDLE;
+		}
+	}
+}
+
+// See stm32f413 programming manual P860
+static inline void I2C_private_ERIRQ(I2C_TypeDef * I2Cx)
+{
+	// Error handling: acknowledge failure
+	if ((I2Cx->SR1 & I2C_SR1_AF) == I2C_SR1_AF)
+	{
+		// Stop current transaction
+		data_length = 0;
+		// Generate a stop condition ONLY ONCE
+		if (I2C_state != I2C_STATE_IDLE){
+			I2C_GenerateSTOP(I2Cx, ENABLE);
+		}
+		// For debugging: toggle the red LED
+		LED_toggleLED(LED_CHANNEL_RED);

```

Comment:
Once you're done testing/debugging, let's replace these LED toggles with a debug_println statement.